rfcs/2023-12-25-proxy-implementation.md

```diff
@@ -0,0 +1,108 @@
+# SMP proxy implementation concerns
+
+* SMP proxy traffic should be indistinguishable from ordinary relay operation.
+* Proxy servers should be implemented by the same binary to prevent maintenance bloat.
+
+This leads to the proxy functionality added directly to the SMP protocol, raising its version.
+
+## Involved parties
+
+### Client
+
+> Client, from the servers' point of view, means the agent running inside a client app. An app/agent protocol is unchanged.
+
+A client that wants to utilize proxy functionality should request protocol version no lower than adds support for it.
+
+A client should be able to configure its usage of SMP proxies.
+Due to proxy/relay integration, there can be some usage patterns (or some blend of them):
+
+1. Single connection for everything. A client uses its only connection for both creating queues for reception and forwarding messages to remote relays. This can be favorable for power-constrained operation or utilizing a TOR hidden service.
+2. Multiple bi-directional SMP connections to public SMP servers. A client spreads its queues over multiple relays as it currently does, and pools them for sending too. Good for redundancy and perhaps sending latency, but also reducing active connections esp. for keen group posters.
+3. Split providers for ingress and egress. Some users may prefer using relays from Provider A, but proxies from Provider B to minimize risks like collusion and correlation.
+
+Option 3 requires either maintaining a proxy list distinct from the relay list or having a "Relay/Proxy/Both" toggle (assuming a server is configured to provide both types).
+
+Additional concern is possibility of using a proxy to forward messages to a relay operating on the same node.
+Can a server refuse to "connect to itself" or should it be client decision entirely?
+- Allowing self-connections may be more efficient as proxy can just short-circuit delivery and this is no worse for privacy than issuing SEND command directly on it.
+- Enforcing this as a general rule may interfere with the scenario 1 from the above list.
+
+Client agent runs a separate "proxy" worker to deal with safety limitations specific to this channel.
+
+When sending a message, a client should make a routing decision.
```

Comment:
This seems not related to the current architecture of the agent/client. This needs to be tied with it. So which part exactly makes this decision. I suspect it most likely will happen in message delivery worker, that will either issue SEND command or provision proxy session.

This whole doc is written as if it will be implemented from scratch, it needs to name the files and functions where things will happen.